jina/types/document/graph.py

```diff
@@ -0,0 +1,322 @@
+from typing import Optional, Iterator, Tuple, Dict, Iterable
+
+import numpy as np
+
+from . import Document, DocumentSourceType
+from ..arrays import ChunkArray
+from ..struct import StructView
+from ..ndarray.sparse.scipy import SparseNdArray
+from ...logging.predefined import default_logger
+
+__all__ = ['GraphDocument']
+
+if False:
+    from scipy.sparse import coo_matrix
+
+
+class GraphDocument(Document):
+    """
+    :class:`GraphDocument` is a data type created based on Jina primitive data type :class:`Document`.
+
+    It adds functionality that lets you work with a `Document` as a `graph` where all its chunks are the nodes in the `graph`.
+
+    It exposes functionality to access and manipulate `graph related info` from the `DocumentProto` such as adjacency and edge features.
+
+    .. warning::
+        - It assumes that every ``chunk`` of a ``document`` is a node of a graph.
+
+    :param document: the document to construct from. If ``bytes`` is given
+            then deserialize a :class:`DocumentProto`; ``dict`` is given then
+            parse a :class:`DocumentProto` from it; ``str`` is given, then consider
+            it as a JSON string and parse a :class:`DocumentProto` from it; finally,
+            one can also give `DocumentProto` directly, then depending on the ``copy``,
+            it builds a view or a copy from it.
+    :param copy: when ``document`` is given as a :class:`DocumentProto` object, build a
+            view (i.e. weak reference) from it or a deep copy from it.
+    :param kwargs: further key value arguments
+    """
+
+    def __init__(
+        self,
+        document: Optional[DocumentSourceType] = None,
+        copy: bool = False,
+        **kwargs,
+    ):
+        super().__init__(document=document, copy=copy, **kwargs)
+        self._node_id_to_offset = {
+            node.id: offset for offset, node in enumerate(self.nodes)
+        }
+
+    def add_node(self, node: 'Document'):
```

Comment:
lets put add methods together and remove methods together, and then properties together